src/psyclone/profiler.py

```diff
@@ -129,6 +129,16 @@ def add_profile_nodes(schedule, loop_class):
                     target = parent_loop
                     parent_loop = parent_loop.ancestor(loop_class)
                 if target:
+                    # Have to take care that the target loop does not have
+                    # a directive applied to it. We distinguish this case
+                    # from that of a directive defining a region by checking
+                    # the number of children of the directive.
+                    if (isinstance(target.parent.parent, Directive) and
+                            len(target.parent.parent.dir_body.children) == 1):
```

Comment:
I am not sure that this test is actually correct - what if the directive has just a single loop (I.e. one child). I modified the nemo test:
```
+++ b/src/psyclone/tests/nemo/transformations/profiling/nemo_profile_test.py
@@ -639,8 +639,10 @@ def test_profile_nemo_acc_kernels(parser):
                                       "  sto_tmp(ji,1) = 0.0d0\n"
                                       "end do\n"
                                       "end program do_loop\n")
-    acctrans.apply(schedule.children)
+    acctrans.apply(schedule.children[0])
+    schedule.view()
     Profiler.add_profile_nodes(schedule, Loop)
+    schedule.view()
     code = str(psy.gen).lower()
     assert ("  type(profile_psydatatype), target, save :: profile_psy_data1\n"
             "  !$acc kernels\n"
(END)
```
I.e. apply the acc trans only to the first loop, not both of them. Then I get:
```
    0: Profile[]
        Schedule[]
            0: Directive[ACC Kernels]
...
    1: Profile[]
        Schedule[]
            0: Loop[type='lon', field_space='None', it_space='None']
```
Now my understanding is that this means we would measure the time on the cpu side, which might not be what we want(?).
Would it be better to have a function in the directive class that indicates if 'code' (can't find the right term atm) can be inserted between it and its child?

Reply:
I think this is OK as I was always imagining that the profiling would be run on the CPU. However, that has made me realise that, as written, this transformation will insert profiling calls *within* an ACC KERNELS region (if it contains >1 loop nest) and that won't compile unless the profiling routines are compiled for the GPU. It's a similar problem to having profiling within an OMP PARALLEL region in that in that case the profiling lib has to be thread safe. I'm not sure what the best solution is. Maybe forbid PSyData nodes from being inside an ACC region?

Reply:
OK, that's the thing I didn't know - I thought GPU profiling would need to be inside OpenACC. It sounds somewhat similar to ``omp do`` - we need to make sure that we don't insert profiling nodes there. So from that point of view I'd say that forbidding profiling nodes after some directives (``omp do``, ``acc``) - though I just realise that this is slightly different: profiling inside an ``omp parallel`` **region** is fine (assuming we link with a thread-safe profiling library), inside an open acc region on the other hand we ... either can't or have to go through a lot of effort.

I'd agree with just disabling it for openacc. We can consider an extension as a separate ticket (e.g. each directive would have a method that specifies what to do with certain PSyData directives - or maybe even more general? ``get_psydata_info()``, which could return 'forbidden', 'up' (move to parent), 'allowed'. That's certainly out of scope for this simple bug fix.

Reply:
While I still don't like the test if there is only one element in a directive, I was not successful in creating a sensible corner case (an ``omp parallel`` containing a non-parallelised loop (e.g. an outer, iterative loop), which then has inside an ``omp do`` parallel loop). This is so hard to construct, that you would need to manually apply the profile transform, and then it's up to you anyway :)